File: src/scripts/order-benchmark.ts
/**
 * Order Placement Benchmark - TypeScript
 *
 * Benchmarks raw order placement latency (sign + post) for $1 orders.
 * Uses pre-cached neg_risk and tickSize to match C++ optimized path.
 *
 * Run: bun run src/scripts/order-benchmark.ts
 *
 * Env:
 *   PRIVATE_KEY - Wallet private key
 *   FUNDER_ADDRESS - Funder address for type-2 signatures
 *   SIGNATURE_TYPE - 1 or 2 (default 2)
 *   NUM_ORDERS - Number of orders to benchmark (default 5)
 */
import { ClobClient, OrderType, Side } from "@polymarket/clob-client";
import { Wallet } from "@ethersproject/wallet";

const privateKey = process.env.PRIVATE_KEY;
const funderAddress = process.env.FUNDER_ADDRESS;
const signatureType = Number(process.env.SIGNATURE_TYPE ?? "2");
const numOrders = Number(process.env.NUM_ORDERS ?? "5");
const CLOB_HOST = "https://clob.polymarket.com";
const CHAIN_ID = 137;

interface BenchmarkResult {
  iteration: number;
  signMs: number;
  postMs: number;
  totalMs: number;
  success: boolean;
  errorMsg?: string;
}

async function main() {
  console.log("=== TypeScript Order Placement Benchmark ===\n");

  if (!privateKey || !funderAddress) {
    console.error("PRIVATE_KEY and FUNDER_ADDRESS required");
    process.exit(1);
  }

  // Initialize client
  const signer = new Wallet(privateKey);
  const temp = new ClobClient(CLOB_HOST, CHAIN_ID, signer);
  let creds;
  try {
    creds = await temp.createOrDeriveApiKey();
  } catch {
    // type-2 may still work
  }
  const client = new ClobClient(CLOB_HOST, CHAIN_ID, signer, creds, signatureType, funderAddress);

  console.log(`Signer: ${signer.address}`);
  console.log(`Funder: ${funderAddress}`);
  console.log(`Orders to benchmark: ${numOrders}\n`);

  // Find active BTC 15m market
  console.log("Finding active BTC 15m market...");
  const now = Math.floor(Date.now() / 1000);
  const currentWindow = Math.floor(now / 900) * 900;

  let tokenId = "";
  let bestAsk = 0;
  let tickSize = "0.01";
  let negRisk = true;

  for (let i = 0; i <= 3; i++) {
    const startTs = currentWindow + i * 900;
    const expTs = startTs + 900;
    if (expTs <= now + 120) continue;

    const slug = `btc-updown-15m-${startTs}`;

    try {
      const response = await fetch(`https://gamma-api.polymarket.com/events?slug=${slug}`);
      const events = (await response.json()) as any[];

      if (!Array.isArray(events) || events.length === 0) continue;

      const event = events[0];
      if (!event.markets || event.markets.length === 0) continue;

      const market = event.markets[0];
      const tokenIds = JSON.parse(market.clobTokenIds);
      tokenId = tokenIds[0]; // YES token

      // Get orderbook for best ask and tick size
      const book = await client.getOrderBook(tokenId);
      if (!book || !(book as any).asks || (book as any).asks.length === 0) continue;

      bestAsk = parseFloat((book as any).asks[0].price);
      tickSize = (book as any).tick_size || "0.01";

      // Get neg_risk (cache it)
      try {
        const negRiskResp = await fetch(`${CLOB_HOST}/neg-risk?token_id=${tokenId}`);
        const negRiskData = (await negRiskResp.json()) as any;
        negRisk = negRiskData.neg_risk ?? true;
      } catch {
        negRisk = true; // Default for crypto markets
      }

      const timeLeft = Math.round((expTs - now) / 60);
      console.log(`Found: ${slug} (expires in ${timeLeft}min)`);
      console.log(`Token: ${tokenId.substring(0, 30)}...`);
      console.log(`Best ask: ${bestAsk}`);
      console.log(`Tick size: ${tickSize}`);
      console.log(`Neg risk: ${negRisk}\n`);
      break;
    } catch {
      continue;
    }
  }

  if (!tokenId) {
    console.error("Could not find active market");
    process.exit(1);
  }

  // Pre-cache order options (this is what TS arb-smoke does)
  const orderOptions = { tickSize: tickSize as any, negRisk };

  // Benchmark loop
  console.log("Starting benchmark...\n");
  console.log("| Iter | Sign (ms) | Post (ms) | Total (ms) | Success |");
  console.log("|------|-----------|-----------|------------|---------|");

  const results: BenchmarkResult[] = [];

  for (let i = 0; i < numOrders; i++) {
    const price = 0.5; // Fixed price for benchmark
    const sizeUsdc = 5.0; // $5 minimum for market orders
    const shares = 10.0; // $5 at 0.50 = 10 shares

    // Time the sign phase
    const signStart = performance.now();
    const order = await client.createOrder(
      {
        tokenID: tokenId,
        price,
        size: shares,
        side: Side.BUY,
      },
      orderOptions
    );
    const signEnd = performance.now();
    const signMs = signEnd - signStart;

    // Time the post phase
    const postStart = performance.now();
    const result = await client.postOrder(order, OrderType.GTC);
    const postEnd = performance.now();
    const postMs = postEnd - postStart;

    const totalMs = signMs + postMs;
    const success = (result as any)?.success === true && !(result as any)?.errorMsg;

    results.push({
      iteration: i + 1,
      signMs,
      postMs,
      totalMs,
      success,
      errorMsg: (result as any)?.errorMsg,
    });

    console.log(
      `| ${(i + 1).toString().padStart(4)} | ${signMs.toFixed(1).padStart(9)} | ${postMs.toFixed(1).padStart(9)} | ${totalMs.toFixed(1).padStart(10)} | ${success ? "  ✓  " : "  ✗  "} |`
    );

    // Small delay between orders
    await new Promise((r) => setTimeout(r, 100));
  }

  // Summary
  console.log("\n=== Summary ===");
  const avgSign = results.reduce((a, b) => a + b.signMs, 0) / results.length;
  const avgPost = results.reduce((a, b) => a + b.postMs, 0) / results.length;
  const avgTotal = results.reduce((a, b) => a + b.totalMs, 0) / results.length;
  const minTotal = Math.min(...results.map((r) => r.totalMs));
  const maxTotal = Math.max(...results.map((r) => r.totalMs));
  const successRate = (results.filter((r) => r.success).length / results.length) * 100;

  console.log(`Average sign time:  ${avgSign.toFixed(1)} ms`);
  console.log(`Average post time:  ${avgPost.toFixed(1)} ms`);
  console.log(`Average total time: ${avgTotal.toFixed(1)} ms`);
  console.log(`Min total time:     ${minTotal.toFixed(1)} ms`);
  console.log(`Max total time:     ${maxTotal.toFixed(1)} ms`);
  console.log(`Success rate:       ${successRate.toFixed(0)}%`);

  // Output JSON for comparison
  console.log("\n=== JSON Output ===");
  console.log(
    JSON.stringify(
      {
        implementation: "typescript",
        numOrders,
        avgSignMs: avgSign,
        avgPostMs: avgPost,
        avgTotalMs: avgTotal,
        minTotalMs: minTotal,
        maxTotalMs: maxTotal,
        successRate,
      },
      null,
      2
    )
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: cpp/src/order_benchmark.cpp
/**
 * Order Placement Benchmark - C++
 *
 * Benchmarks raw order placement latency (sign + post) for $1 orders.
 * Uses pre-cached neg_risk to skip API calls during order creation.
 *
 * Build: cmake --build build --target order_benchmark
 * Run: PRIVATE_KEY=0x... FUNDER_ADDRESS=0x... ./build/order_benchmark
 *
 * Env:
 *   PRIVATE_KEY      - Wallet private key (required)
 *   FUNDER_ADDRESS   - Address holding funds (for proxy wallets)
 *   NUM_ORDERS       - Number of orders to benchmark (default: 5)
 */

#include "order_signer.hpp"
#include "http_client.hpp"
#include <nlohmann/json.hpp>
#include <iostream>
#include <cstdlib>
#include <chrono>
#include <iomanip>
#include <vector>
#include <numeric>
#include <algorithm>
#include <thread>

using json = nlohmann::json;
using namespace polymarket;

const std::string CLOB_API = "https://clob.polymarket.com";
const std::string NEG_RISK_CTF_EXCHANGE = "0xC5d563A36AE78145C45a50134d48A1215220f80a";
const std::string CTF_EXCHANGE = "0x4bFb41d5B3570DeFd03C39a9A4D8dE6Bd8B8982E";

struct BenchmarkResult
{
    int iteration;
    double sign_ms;
    double post_ms;
    double total_ms;
    bool success;
    std::string error_msg;
};

int main(int argc, char *argv[])
{
    const char *private_key_env = std::getenv("PRIVATE_KEY");
    const char *funder_address_env = std::getenv("FUNDER_ADDRESS");
    const char *num_orders_env = std::getenv("NUM_ORDERS");

    if (!private_key_env)
    {
        std::cerr << "Error: PRIVATE_KEY environment variable required\n";
        return 1;
    }

    std::string private_key = private_key_env;
    std::string funder_address = funder_address_env ? funder_address_env : "";
    int num_orders = num_orders_env ? std::stoi(num_orders_env) : 5;

    std::cout << "=== C++ Order Placement Benchmark ===\n\n";

    // Initialize signer
    OrderSigner signer(private_key);
    if (funder_address.empty())
    {
        funder_address = signer.address();
    }

    std::cout << "Signer: " << signer.address() << "\n";
    std::cout << "Funder: " << funder_address << "\n";
    std::cout << "Orders to benchmark: " << num_orders << "\n\n";

    http_global_init();
    HttpClient http;
    http.set_base_url(CLOB_API);
    http.set_timeout_ms(10000);

    // Derive API credentials
    std::cout << "Deriving API credentials...\n";
    ApiCredentials creds;
    try
    {
        creds = signer.create_or_derive_api_credentials(http, funder_address);
        std::cout << "API key: " << creds.api_key.substr(0, 8) << "...\n\n";
    }
    catch (const std::exception &e)
    {
        std::cerr << "Failed: " << e.what() << "\n";
        http_global_cleanup();
        return 1;
    }

    // Find active BTC 15m market
    std::cout << "Finding active BTC 15m market...\n";

    uint64_t now_ts = static_cast<uint64_t>(std::time(nullptr));
    uint64_t current_window = (now_ts / 900) * 900;

    std::string token_id;
    double best_ask = 0;
    bool is_neg_risk = false;
    std::string exchange_address;

    for (int i = 0; i <= 3; i++)
    {
        uint64_t start_ts = current_window + i * 900;
        uint64_t exp_ts = start_ts + 900;
        if (exp_ts <= now_ts + 120)
            continue;

        std::string slug = "btc-updown-15m-" + std::to_string(start_ts);

        HttpClient gamma_http;
        gamma_http.set_base_url("https://gamma-api.polymarket.com");
        gamma_http.set_timeout_ms(10000);

        auto gamma_response = gamma_http.get("/events?slug=" + slug);
        if (!gamma_response.ok())
            continue;

        try
        {
            auto gamma_json = json::parse(gamma_response.body);
            if (!gamma_json.is_array() || gamma_json.empty())
                continue;

            auto &event = gamma_json[0];
            if (!event.contains("markets") || event["markets"].empty())
                continue;

            auto &mkt = event["markets"][0];
            auto token_ids = json::parse(mkt["clobTokenIds"].get<std::string>());
            token_id = token_ids[0].get<std::string>(); // YES token

            // Get orderbook for best ask
            auto book_response = http.get("/book?token_id=" + token_id);
            if (!book_response.ok())
                continue;

            auto book_json = json::parse(book_response.body);
            if (!book_json.contains("asks") || book_json["asks"].empty())
                continue;

            best_ask = std::stod(book_json["asks"][0]["price"].get<std::string>());

            // Get neg_risk (cache it once)
            auto neg_risk_response = http.get("/neg-risk?token_id=" + token_id);
            if (neg_risk_response.ok())
            {
                auto neg_risk_json = json::parse(neg_risk_response.body);
                is_neg_risk = neg_risk_json.value("neg_risk", false);
            }

            exchange_address = is_neg_risk ? NEG_RISK_CTF_EXCHANGE : CTF_EXCHANGE;

            uint64_t time_left = (exp_ts - now_ts) / 60;
            std::cout << "Found: " << slug << " (expires in " << time_left << "min)\n";
            std::cout << "Token: " << token_id.substr(0, 30) << "...\n";
            std::cout << "Best ask: " << best_ask << "\n";
            std::cout << "Neg risk: " << (is_neg_risk ? "true" : "false") << "\n\n";
            break;
        }
        catch (...)
        {
            continue;
        }
    }

    if (token_id.empty())
    {
        std::cerr << "Could not find active market\n";
        http_global_cleanup();
        return 1;
    }

    // Benchmark loop
    std::cout << "Starting benchmark...\n\n";
    std::cout << "| Iter | Sign (ms) | Post (ms) | Total (ms) | Success |\n";
    std::cout << "|------|-----------|-----------|------------|---------|" << std::endl;

    std::vector<BenchmarkResult> results;

    for (int i = 0; i < num_orders; i++)
    {
        double price = 0.50;    // Fixed price for benchmark
        double size_usdc = 5.0; // $5 minimum for market orders
        double shares = 10.0;   // $5 at 0.50 = 10 shares

        // Time the sign phase
        auto sign_start = std::chrono::high_resolution_clock::now();

        OrderData order_data;
        order_data.maker = funder_address;
        order_data.taker = "0x0000000000000000000000000000000000000000";
        order_data.token_id = token_id;
        order_data.maker_amount = to_wei(size_usdc, 6);
        order_data.taker_amount = to_wei(shares, 6);
        order_data.side = OrderSide::BUY;
        order_data.fee_rate_bps = "0";
        order_data.nonce = "0";
        order_data.signer = signer.address();
        order_data.expiration = "0";
        order_data.signature_type = (funder_address != signer.address())
                                        ? SignatureType::POLY_GNOSIS_SAFE
                                        : SignatureType::EOA;

        // Use cached exchange_address (no API call!)
        auto signed_order = signer.sign_order(order_data, exchange_address);

        auto sign_end = std::chrono::high_resolution_clock::now();
        double sign_ms = std::chrono::duration<double, std::milli>(sign_end - sign_start).count();

        // Build order payload
        nlohmann::ordered_json payload;
        nlohmann::ordered_json order_obj;
        order_obj["salt"] = std::stoll(signed_order.salt);
        order_obj["maker"] = signed_order.maker;
        order_obj["signer"] = signed_order.signer;
        order_obj["taker"] = signed_order.taker;
        order_obj["tokenId"] = signed_order.token_id;
        order_obj["makerAmount"] = signed_order.maker_amount;
        order_obj["takerAmount"] = signed_order.taker_amount;
        order_obj["side"] = "BUY";
        order_obj["expiration"] = signed_order.expiration;
        order_obj["nonce"] = signed_order.nonce;
        order_obj["feeRateBps"] = signed_order.fee_rate_bps;
        order_obj["signatureType"] = static_cast<int>(signed_order.signature_type);
        order_obj["signature"] = signed_order.signature;
        payload["order"] = order_obj;
        payload["orderType"] = "GTC"; // Use GTC so order posts even without liquidity match
        payload["owner"] = creds.api_key;

        std::string body_str = payload.dump();

        // Generate L2 headers
        auto l2_hdrs = signer.generate_l2_headers(creds, "POST", "/order", body_str, funder_address);
        std::map<std::string, std::string> headers = {
            {"POLY_ADDRESS", l2_hdrs.poly_address},
            {"POLY_SIGNATURE", l2_hdrs.poly_signature},
            {"POLY_TIMESTAMP", l2_hdrs.poly_timestamp},
            {"POLY_API_KEY", l2_hdrs.poly_api_key},
            {"POLY_PASSPHRASE", l2_hdrs.poly_passphrase}};

        // Time the post phase
        auto post_start = std::chrono::high_resolution_clock::now();
        auto response = http.post("/order", body_str, headers);
        auto post_end = std::chrono::high_resolution_clock::now();
        double post_ms = std::chrono::duration<double, std::milli>(post_end - post_start).count();

        double total_ms = sign_ms + post_ms;

        bool success = false;
        std::string error_msg;

        if (response.ok())
        {
            try
            {
                auto result = json::parse(response.body);
                success = result.value("success", false) && result.value("errorMsg", "") == "";
                error_msg = result.value("errorMsg", "");
                if (!success && error_msg.empty())
                {
                    error_msg = response.body.substr(0, 100);
                }
            }
            catch (...)
            {
                error_msg = "parse error: " + response.body.substr(0, 100);
            }
        }
        else
        {
            error_msg = "HTTP " + std::to_string(response.status_code) + ": " + response.body.substr(0, 100);
        }

        // Debug: print error for first order
        if (i == 0 && !success)
        {
            std::cerr << "DEBUG first order error: " << error_msg << "\n\n";
        }

        results.push_back({i + 1, sign_ms, post_ms, total_ms, success, error_msg});

        std::cout << "| " << std::setw(4) << (i + 1)
                  << " | " << std::setw(9) << std::fixed << std::setprecision(1) << sign_ms
                  << " | " << std::setw(9) << post_ms
                  << " | " << std::setw(10) << total_ms
                  << " | " << (success ? "  ✓  " : "  ✗  ") << " |" << std::endl;

        // Small delay between orders
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    // Summary
    std::cout << "\n=== Summary ===\n";

    double avg_sign = 0, avg_post = 0, avg_total = 0;
    double min_total = results[0].total_ms, max_total = results[0].total_ms;
    int success_count = 0;

    for (const auto &r : results)
    {
        avg_sign += r.sign_ms;
        avg_post += r.post_ms;
        avg_total += r.total_ms;
        min_total = std::min(min_total, r.total_ms);
        max_total = std::max(max_total, r.total_ms);
        if (r.success)
            success_count++;
    }

    avg_sign /= results.size();
    avg_post /= results.size();
    avg_total /= results.size();
    double success_rate = (static_cast<double>(success_count) / results.size()) * 100;

    std::cout << "Average sign time:  " << std::fixed << std::setprecision(1) << avg_sign << " ms\n";
    std::cout << "Average post time:  " << avg_post << " ms\n";
    std::cout << "Average total time: " << avg_total << " ms\n";
    std::cout << "Min total time:     " << min_total << " ms\n";
    std::cout << "Max total time:     " << max_total << " ms\n";
    std::cout << "Success rate:       " << std::setprecision(0) << success_rate << "%\n";

    // Output JSON for comparison
    std::cout << "\n=== JSON Output ===\n";
    json output;
    output["implementation"] = "cpp";
    output["numOrders"] = num_orders;
    output["avgSignMs"] = avg_sign;
    output["avgPostMs"] = avg_post;
    output["avgTotalMs"] = avg_total;
    output["minTotalMs"] = min_total;
    output["maxTotalMs"] = max_total;
    output["successRate"] = success_rate;
    std::cout << output.dump(2) << std::endl;

    http_global_cleanup();
    return 0;
}
